### PR TITLE
LDAP alias provisioning

### LINKED ISSUE #5173: Provision aliases based on an LDAP attribute
### Feature Request

Allows admins to optionally configure aliases for provisioned accounts. Since LDAP is already a popular choice for a user back-end of Nextcloud, we can build this feature around LDAP. The admins will have to enter the attribute name and we use ``\OCP\LDAP\ILDAPProvider::getUserAttribute`` to fetch the alias email address.

If LDAP isn't available (`\OCP\LDAP\ILDAPProviderFactory::isAvailable`) then we should hide this setting from the admin.

The attribute could be multivalued if more than one alias is set. We have to check if/how `\OCP\LDAP\ILDAPProvider::getUserAttribute` supports this because it currently always returns a string. Taking a quick look at the implementation tells me only the first result will be returned if the value is an array. This means we might have to change the server API to allow multiple values, e.g. in a `getUserAttributes` or `getUserMultiAttribute`

#### Summary

Provision aliases based on LDAP attributes.

### PR DESCRIPTION
**Pull request by kesselb**
Fix #5173

- [x] Requires: https://github.com/nextcloud/server/pull/27515. 

**Todo**

- [x] Add option to enable/disable alias provisioning
- [x] Add option to customize the ldap attribute
- [ ] ~Add option to filter aliases by regex~
- [x] Alias section is currently hidden for provisioned accounts. Show the section but disable add alias / delete alias
- [x] It should be possible to change the name but not the email/alias
- [x] Test that editing the signature for alias works
- [x] Add a database field (or virtual field) to alias if provisioned or not
- [ ] ~Add unique constraint to mail_aliases account_id and alias.~
- [x] UI to enable/disable, ldap attribute, set filter.

### CODE REVIEW

File: lib/Db/Alias.php [hunk 1]
```diff
@@ -63,6 +68,7 @@ public function jsonSerialize(): array {
 			'name' => $this->getName(),
 			'alias' => $this->getAlias(),
 			'signature' => $this->getSignature(),
+			'provisioned' => (bool)$this->getProvisioningId(),
```
Comment (ChristophWurst) [inline lib/Db/Alias.php:71]:
```suggestion
			'provisioned' => $this->getProvisioningId() !== null,
```

maybe the bool cast works too but :see_no_evil:

File: src/components/AliasForm.vue [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+				@click="updateAlias" />
+		</form>
+		<span v-else>
+			<strong>{{ alias.name }}</strong> &lt;{{ alias.alias }}&gt;
+			<button
+				class="icon icon-rename"
+				:title="t('mail', 'Show update alias form')"
+				@click="showForm = true" />
+			<button v-if="!alias.provisioned"
+				class="icon icon-delete"
```
Comment (kesselb) [inline src/components/AliasForm.vue:40]: loading state for delete is missing

File: lib/Service/Provisioning/Manager.php [hunk 3]
```diff
@@ -84,6 +98,71 @@ public function provision(): int {
...
+
+	/**
+	 * @throws \Exception if user id was not found in LDAP
+	 */
+	public function ldapAliasesIntegration(Provisioning $provisioning, IUser $user): Provisioning {
+		if ($user->getBackendClassName() !== 'LDAP' || $provisioning->getLdapAliasesProvisioning() === false || empty($provisioning->getLdapAliasesAttribute())) {
+			return $provisioning;
+		}
+
+		if ($this->ldapProviderFactory->isAvailable() === false) {
```
Comment (kesselb) [inline lib/Service/Provisioning/Manager.php:155]: Snap! isAvailable is new in 21 :man_shrugging:

Comment (ChristophWurst) [reply lib/Service/Provisioning/Manager.php:158]: Do an additional server version comparison here and don't offer the feature for Nextcloud 20 or older. It's a fair tradeoff IMO.

Comment (kesselb) [reply lib/Service/Provisioning/Manager.php:158]: Thanks! I added `@psalm-suppress UndefinedInterfaceMethod` to make Psalm on Nextcloud 20 happy. The check for the right Nextcloud version is already done some lines earlier: https://github.com/nextcloud/mail/pull/5198/files#diff-fe646e6e2f7561cf6a8731328355c7432be7647f7114e1619fe2bf5ff2638c42R205. If `getMultiValueUserAttribute` exist `isAvailable` also exist.

File: lib/Controller/SettingsController.php [hunk 4]
```diff
@@ -60,19 +60,20 @@ public function createProvisioning(array $data): JSONResponse {
 			$this->provisioningManager->newProvisioning($data);
 		} catch (ValidationException $e) {
 			return HttpJsonResponse::fail([$e->getFields()]);
+		} catch (\Exception $e) {
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php:63]: could you elaborate why this was widened?

File: lib/Controller/SettingsController.php [hunk 5]
```diff
@@ -60,19 +60,20 @@ public function createProvisioning(array $data): JSONResponse {
...
 	public function updateProvisioning(int $id, array $data): JSONResponse {
 		try {
-			$this->provisioningManager->updateProvisioning(array_merge(
-				$data,
-				['id' => $id]
-			));
+			$this->provisioningManager->updateProvisioning(array_merge($data, ['id' => $id]));
 		} catch (ValidationException $e) {
 			return HttpJsonResponse::fail([$e->getFields()]);
+		} catch (\Exception $e) {
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php:75]: same. Could this possibly be a ClientException? \Error will throw almost anything. That *anything* could also be from unexpected service errors, then the HTTP4xx isn't appropriate IMO

File: lib/Db/AliasMapper.php [hunk 6]
```diff
@@ -89,6 +107,29 @@ public function deleteAll($accountId) {
 		$query->execute();
 	}
 
+	/**
+	 * Delete all provisioned aliases for the given uid
+	 *
+	 * Exception for Nextcloud 20: \Doctrine\DBAL\DBALException
+	 * Exception for Nextcloud 21 and newer: \OCP\DB\Exception
+	 *
+	 * @TODO: Change throws to \OCP\DB\Exception once Mail does not support Nextcloud 20.
```
Comment (ChristophWurst) [inline lib/Db/AliasMapper.php:116]: FYI I think we do this soon-ish so we only have to support two major versions

**Review by ChristophWurst — COMMENTED** [review]
Code looks good otherwise :)

File: lib/Controller/SettingsController.php (hunk 5 above)
Comment (kesselb) [reply lib/Controller/SettingsController.php:75]: `$this->provisioningManager->newProvisioning($data);` throws a validation exception if input data is invalid or a exception related to the database. I guess we can migrate this to Throwable and `HttpJsonResponse.errorFromThrowable`.

File: lib/Service/Provisioning/Manager.php [hunk 7]
```diff
@@ -84,6 +98,75 @@ public function provision(): int {
...
+
+		/** @psalm-suppress UndefinedInterfaceMethod */
+		if ($this->ldapProviderFactory->isAvailable() === false) {
+			$this->logger->debug('Request to provision mail aliases but ldap not available');
+			return $provisioning;
+		}
+
+		$ldapProvider = $this->ldapProviderFactory->getLDAPProvider();
+		/** @psalm-suppress UndefinedInterfaceMethod */
+		$provisioning->setAliases($ldapProvider->getMultiValueUserAttribute($user->getUID(), $provisioning->getLdapAliasesAttribute()));
```
Comment (kesselb) [inline lib/Service/Provisioning/Manager.php:165]: @kesselb add validation for data

**Review by st3iny — APPROVED** [review]
Setup a test environment using your guide and tested it locally with multiple aliases. Works fine.

The only issue I noticed it that once an account has been provisioned, changes of the aliases in LDAP won't be forwarded to the mail alias settings. 

E.g. If I remove the alias from LDAP and then run the provisioning again, the old alias is still configured in mail.

**Comment by kesselb** [comment]
> Setup a test environment using your guide and tested it locally with multiple aliases. Works fine.
> 
> The only issue I noticed it that once an account has been provisioned, changes of the aliases in LDAP won't be forwarded to the mail alias settings.
> 
> E.g. If I remove one of multiple aliases from LDAP and then run the provisioning again, all aliases are still configured in mail.

That should work.

**Comment by kesselb** [comment]
> > Setup a test environment using your guide and tested it locally with multiple aliases. Works fine.
> > The only issue I noticed it that once an account has been provisioned, changes of the aliases in LDAP won't be forwarded to the mail alias settings.
> > E.g. If I remove one of multiple aliases from LDAP and then run the provisioning again, all aliases are still configured in mail.
> 
> That should work.

The provisioning run with every http request in mail. To reduce the load to the ldap server the response for getMultiValueUserAttribute is cached (in server). 
If a distributed cache is configured the value is cached. You may set ldapCacheTTL to 0 for testing.

**Comment by st3iny** [comment]
> If a distributed cache is configured the value is cached. You may set ldapCacheTTL to 0 for testing.

That has been the issue. My dev setup has a distributed cache setup. Thanks for clarifying :)

**Pull request merged**